File: packages/r/tests/testthat/test-api.R
reseal_index <- function(index, cache_dir) {
  text <- jsonlite::toJSON(
    index,
    auto_unbox = TRUE,
    null = "null",
    digits = NA,
    pretty = FALSE
  )
  contents <- charToRaw(text)
  digest <- digest::digest(contents, algo = "sha256", serialize = FALSE)
  directory <- file.path(cache_dir, "registries", "sha256")
  dir.create(directory, recursive = TRUE, showWarnings = FALSE)
  writeBin(contents, file.path(directory, digest))
  registry_selector(
    index$release,
    digest,
    "https://example.com/index.json"
  )
}

test_that("metadata operations expose identity and provenance without artifacts", {
  cache_dir <- tempfile("datamonger-cache-")
  seeded <- seed_conformance_cache(cache_dir)

  info <- data_info(
    "mixed_csv",
    source = "conformance",
    registry = seeded$selector,
    cache_dir = cache_dir,
    offline = TRUE
  )
  listed <- list_data(
    registry = seeded$selector,
    cache_dir = cache_dir,
    offline = TRUE
  )

  expect_s3_class(info, "datamonger_data_info")
  expect_identical(info$dataset_id, "conformance:mixed_csv@1")
  expect_true(nzchar(info$provenance$provider))
  expect_length(listed, 5L)
  expect_true(all(vapply(listed, inherits, logical(1), "datamonger_data_info")))
})

test_that("decoded verification can be disabled only explicitly", {
  cache_dir <- tempfile("datamonger-cache-")
  seeded <- seed_conformance_cache(cache_dir)
  result <- fetch_data(
    "mixed_csv",
    source = "conformance",
    registry = seeded$selector,
    cache_dir = cache_dir,
    offline = TRUE,
    verify_decoded = FALSE,
    return_info = TRUE
  )

  expect_identical(result$info$verification, "artifact")
  expect_null(result$info$canonical_form)
  expect_null(result$info$canonical_digest)
  expect_error(
    fetch_data(
      "mixed_csv",
      source = "conformance",
      registry = seeded$selector,
      cache_dir = cache_dir,
      offline = TRUE,
      verify_decoded = NA
    ),
    "must be TRUE or FALSE"
  )
})

test_that("public failures use the shared semantic taxonomy", {
  cache_dir <- tempfile("datamonger-cache-")
  seeded <- seed_conformance_cache(cache_dir)
  expect_error(
    data_info(
      "missing",
      source = "conformance",
      registry = seeded$selector,
      cache_dir = cache_dir,
      offline = TRUE
    ),
    class = "datamonger_unknown_dataset"
  )

  index <- seeded$index
  index$datasets[[1]]$representation$decoder_version <- 2L
  unsupported <- reseal_index(index, cache_dir)
  expect_error(
    fetch_data(
      "mixed_csv",
      source = "conformance",
      registry = unsupported,
      cache_dir = cache_dir,
      offline = TRUE
    ),
    class = "datamonger_unsupported_decoder"
  )

  index <- seeded$index
  index$datasets[[1]]$artifacts[[1]]$distribution <- "metadata-only"
  index$datasets[[1]]$artifacts[[1]]$downloads <- NULL
  metadata_only <- reseal_index(index, cache_dir)
  expect_error(
    fetch_artifact(
      "mixed_csv",
      source = "conformance",
      registry = metadata_only,
      cache_dir = cache_dir,
      offline = TRUE
    ),
    class = "datamonger_artifact_unavailable"
  )

  index <- seeded$index
  index$datasets[[1]]$representation$expect$verification[[1]]$digest <- paste(rep("0", 64), collapse = "")
  mismatch <- reseal_index(index, cache_dir)
  expect_error(
    fetch_data(
      "mixed_csv",
      source = "conformance",
      registry = mismatch,
      cache_dir = cache_dir,
      offline = TRUE
    ),
    class = "datamonger_decoded_integrity"
  )
})

test_that("verification errata revoke records independent of JSON key order", {
  cache_dir <- tempfile("datamonger-cache-")
  seeded <- seed_conformance_cache(cache_dir)
  index <- seeded$index
  dataset <- index$datasets[[1]]
  replacement <- dataset$representation$expect$verification[[1]]
  original <- list(
    canonical_form = 1L,
    algorithm = "sha256",
    digest = paste(rep("0", 64), collapse = "")
  )
  dataset$representation$expect$verification <- list(
    original,
    replacement
  )
  index$datasets[[1]] <- dataset
  index$errata <- list(list(
    schema_version = 1L,
    id = "mixed-csv-verification",
    release = "test-0001",
    dataset = list(
      name = dataset$name,
      version = dataset$version,
      source = dataset$source
    ),
    target = list(
      kind = "verification",
      canonical_form = 1L,
      algorithm = "sha256"
    ),
    original = list(
      algorithm = "sha256",
      digest = original$digest,
      canonical_form = 1L
    ),
    replacement = replacement,
    reason = "The earlier digest was incorrect.",
    approval = list(maintainer = "fixture", approved_at = "2026-09-07")
  ))
  selected <- reseal_index(index, cache_dir)

  result <- fetch_data(
    dataset$name,
    source = dataset$source,
    version = dataset$version,
    registry = selected,
    cache_dir = cache_dir,
    offline = TRUE,
    return_info = TRUE
  )

  expect_identical(result$info$canonical_digest, replacement$digest)
})

File: packages/julia/test/test_api.jl
@testset "public API conformance" begin
    cache_dir = mktempdir()
    seeded = seed_conformance_cache(cache_dir)

    info = data_info(
        "mixed_csv";
        source="conformance",
        registry=seeded.selector,
        cache_dir,
        offline=true,
    )
    @test info.dataset_id == "conformance:mixed_csv@1"
    @test !isempty(info.provenance["provider"])
    @test length(list_data(; registry=seeded.selector, cache_dir, offline=true)) == 5

    table = fetch_data(
        "mixed_csv";
        source="conformance",
        registry=seeded.selector,
        cache_dir,
        offline=true,
    )
    @test table isa DataFrame
    @test size(table) == (5, 4)
    @test table[2, :label] == "quoted, value"
    @test table[3, :label] == "says \"hello\""
    @test ismissing(table[3, :measurement])
    @test eltype(table.measurement) == Union{Missing,Float64}
    @test eltype(table.count) == Union{Missing,Int64}

    cases = read_json("conformance", "cases.json")["cases"]
    expected = Dict(case["dataset"] => case["expected_sha256"] for case in cases)
    for dataset in seeded.index["datasets"]
        result = fetch_data(
            dataset["name"];
            source=dataset["source"],
            version=dataset["version"],
            registry=seeded.selector,
            cache_dir,
            offline=true,
            return_info=true,
        )
        dataset_id = "$(dataset["source"]):$(dataset["name"])@$(dataset["version"])"
        @test result.info.verification == :decoded
        @test result.info.canonical_digest == expected[dataset_id]
    end

    result = fetch_data(
        "mixed_csv";
        source="conformance",
        registry=seeded.selector,
        cache_dir,
        offline=true,
        verify_decoded=false,
        return_info=true,
    )
    @test result.info.verification == :artifact
    @test isnothing(result.info.canonical_form)
    @test_throws UnknownDatasetError data_info(
        "missing";
        source="conformance",
        registry=seeded.selector,
        cache_dir,
        offline=true,
    )

    @test_throws ArtifactSelectionError fetch_artifact(
        "small_libsvm_split";
        source="conformance",
        registry=seeded.selector,
        cache_dir,
        offline=true,
    )
end

@testset "public semantic failures and errata" begin
    cache_dir = mktempdir()
    seeded = seed_conformance_cache(cache_dir)

    index = deepcopy(seeded.index)
    index["schema_version"] = 2
    unsupported_registry = reseal_index(index, cache_dir)
    @test_throws UnsupportedRegistryError list_data(
        ; registry=unsupported_registry, cache_dir, offline=true,
    )

    index = deepcopy(seeded.index)
    index["datasets"][1]["representation"]["decoder_version"] = 2
    unsupported_decoder = reseal_index(index, cache_dir)
    @test_throws UnsupportedDecoderError fetch_data(
        "mixed_csv";
        source="conformance",
        registry=unsupported_decoder,
        cache_dir,
        offline=true,
    )

    index = deepcopy(seeded.index)
    artifact = index["datasets"][1]["artifacts"][1]
    artifact["distribution"] = "metadata-only"
    delete!(artifact, "downloads")
    metadata_only = reseal_index(index, cache_dir)
    @test_throws ArtifactUnavailableError fetch_artifact(
        "mixed_csv";
        source="conformance",
        registry=metadata_only,
        cache_dir,
        offline=true,
    )

    index = deepcopy(seeded.index)
    index["datasets"][1]["representation"]["expect"]["verification"][1]["digest"] =
        repeat("0", 64)
    mismatch = reseal_index(index, cache_dir)
    @test_throws DecodedIntegrityError fetch_data(
        "mixed_csv";
        source="conformance",
        registry=mismatch,
        cache_dir,
        offline=true,
    )

    index = deepcopy(seeded.index)
    dataset = index["datasets"][1]
    replacement = deepcopy(dataset["representation"]["expect"]["verification"][1])
    original = Dict(
        "canonical_form" => 1,
        "algorithm" => "sha256",
        "digest" => repeat("0", 64),
    )
    dataset["representation"]["expect"]["verification"] = Any[original, replacement]
    index["errata"] = Any[Dict(
        "schema_version" => 1,
        "id" => "mixed-csv-verification",
        "release" => index["release"],
        "dataset" => Dict(
            "name" => dataset["name"],
            "version" => dataset["version"],
            "source" => dataset["source"],
        ),
        "target" => Dict(
            "kind" => "verification",
            "canonical_form" => 1,
            "algorithm" => "sha256",
        ),
        "original" => Dict(
            "algorithm" => original["algorithm"],
            "digest" => original["digest"],
            "canonical_form" => original["canonical_form"],
        ),
        "replacement" => replacement,
        "reason" => "The earlier digest was incorrect.",
        "approval" => Dict(
            "maintainer" => "fixture",
            "approved_at" => "2026-09-07",
        ),
    )]
    corrected = reseal_index(index, cache_dir)
    result = fetch_data(
        dataset["name"];
        source=dataset["source"],
        version=dataset["version"],
        registry=corrected,
        cache_dir,
        offline=true,
        return_info=true,
    )
    @test result.info.canonical_digest == replacement["digest"]
end

@testset "semantic taxonomy is complete" begin
    expected = Set(case["expected"] for case in read_json("conformance", "errors.json")["cases"])
    @test Set(keys(Datamonger.SEMANTIC_ERROR_CATEGORIES)) == expected
end
